### PR TITLE
Fix transaction performance collector oom

### PR DESCRIPTION
**Pull request by stefanosiano**
## :scroll: Description
PerformanceCollectionData now contains single values instead of lists
ICollectors now work on a single instance of PerformanceCollectionData instead than a list
TransactionPerformanceCollector:
* is now set in Android options
* a NoOpTransactionPerformanceCollector is set in Java options
* deletes data after 30 seconds
* collect data only if a profile is sampled

## :bulb: Motivation and Context
We collect data about cpu and memory usage during transactions. This caused [oom error in java backend](https://github.com/getsentry/sentry-java/issues/2496).  
As a workaround, the data collection [was disabled](https://github.com/getsentry/sentry-java/pull/2498).   
We are now enabling the data collection only on the Android SDK, and only if a profile is sampled, since we send this data only in profiles, at the moment.
Several updates were put to reduce the memory footprint of the collected data

## :green_heart: How did you test it?
Unit test

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

**Review by markushi — COMMENTED** [review]
Looks good, please have a look at my comments!

File: sentry/src/main/java/io/sentry/DefaultTransactionPerformanceCollector.java [hunk 1]
```diff
@@ -74,40 +69,51 @@ public void run() {
...
-                    data.commitData();
-                  }
+                final @NotNull PerformanceCollectionData tempData = new PerformanceCollectionData();
+
+                for (ICollector collector : collectors) {
+                  collector.collect(tempData);
+                }
+                // We commit data after calling all collectors.
+                // This way we avoid issues caused by having multiple cpu or memory collectors.
+                tempData.commitData();
```
[the remark this answers is not in the document]
Comment (stefanosiano) [reply sentry/src/main/java/io/sentry/DefaultTransactionPerformanceCollector.java:83]: Yeah, the thing was not to have issues due to multiple memory collectors.
Previously I was using a list, and that commit was needed, but i can safely remove that now.

File: sentry/src/main/java/io/sentry/DefaultTransactionPerformanceCollector.java [hunk 2]
```diff
@@ -74,40 +69,51 @@ public void run() {
...
+
+                for (List<PerformanceCollectionData> data : performanceDataMap.values()) {
```
[the remark this answers is not in the document]
Comment (stefanosiano) [reply sentry/src/main/java/io/sentry/DefaultTransactionPerformanceCollector.java:82]: performanceDataMap is a `ConcurrentHashMap`, so no concurrentModifications occur.
I previously put a synchronized block, but that caused problems when there are a lot of transactions (100k+). It's an edge case, and only for backend, which would have performance collection disabled anyway, but I removed that and relied on the `ConcurrentHashMap`

**Comment by adinauer** [comment]
Let me know if you need my help in testing if OOM is solved.

**Review by markushi — APPROVED** [review]
LGTM!

**Comment by stefanosiano** [comment]
> Let me know if you need my help in testing if OOM is solved.

Thanks @adinauer! I tried with your commands, and i found that after forcing a gc the ram went back to normal.
Feel free to try, just remember that it's still disabled for the backend. So you'd need to set the `DefaultTransactionPerformanceCollector` in the options and remove the `&& Boolean.TRUE.equals(isProfileSampled())` from [this line](https://github.com/getsentry/sentry-java/blob/fix/transaction-performance-collector-oom-fix/sentry/src/main/java/io/sentry/SentryTracer.java#L146)

**Pull request merged**